[Merged by Bors] - feat(algebra/algebra/spectrum): lemmas when scalars are a field

j-loreaux opened the pull request:
Prove some properties of the spectrum when the underlying scalar ring
is a field, and mostly assuming the algebra is itself nontrivial.
Show that the spectrum of a scalar (i.e., `algebra_map 𝕜 A k`) is
the singleton `{k}`. Prove that `σ (a * b) \ {0} = σ (b * a) \ {0}`.



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux committed:
feat(algebra/algebra/spectrum): lemmas when scalars are a field

Prove some properties of the spectrum when the underlying scalar ring
is a field, and mostly assuming the algebra is itself nontrivial.
Show that the spectrum of a scalar (i.e., `algebra_map 𝕜 A k`) is
the singleton `{k}`. Prove that `σ (a * b) \ {0} = σ (b * a) \ {0}`.

ericrbg commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -167,4 +170,69 @@ theorem preimage_units_mul_eq_swap_mul {a b : A} :
+    rw mem_resolvent_iff,
+    simpa [algebra.algebra_map_eq_smul_one], },
+  { rw [set.singleton_subset_iff, mem_iff],
+    simp [algebra.algebra_map_eq_smul_one], },
```
```suggestion
  refine set.eq_of_subset_of_subset _ (by simp [algebra.algebra_map_eq_smul_one, mem_iff]),
  rw [spectrum, set.compl_subset_comm],
  intros k hk,
  rw set.mem_compl_singleton_iff at hk,
  have : is_unit (units.mk0 k hk • (1 : A)) := is_unit.smul (units.mk0 k hk) is_unit_one,
  simpa [mem_resolvent_iff, algebra.algebra_map_eq_smul_one]
```

🏌️, but more importantly I think it's advocated to avoid `change` when possible, as it can be a bit brittle. also, I think the more common spelling of that first lemma is `set.subset.antisymm`, but it's fine either way

j-loreaux replied:
good to know about `change`

ericrbg commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -167,4 +170,69 @@ theorem preimage_units_mul_eq_swap_mul {a b : A} :
+  σ (k • a) = k • (σ a) :=
+begin
+  by_cases h : k = 0,
+  { simp [h,ha,zero_smul_set], refl },
```
```suggestion
  rcases eq_or_ne k 0 with rfl | h,
  { simpa [ha, zero_smul_set] },
```

nonterminal simps be bad

j-loreaux replied:
good to know, thanks.

j-loreaux replied:
That `rfl` syntax for `rcases` is really nice. I hadn't known about that before.

ericrbg commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -167,4 +170,69 @@ theorem preimage_units_mul_eq_swap_mul {a b : A} :
+
+open_locale pointwise
+
+/-- the assumption (σ a).nonempty is necessary and cannot be removed without
```
what are these conditions? anything nice like `no_zero_smul_divisors`?

j-loreaux replied:
no, basically the only case I know where it's guaranteed is where the scalars are the complex field and the algebra is a Banach algebra.

ericrbg commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -167,4 +170,69 @@ theorem preimage_units_mul_eq_swap_mul {a b : A} :
+
+theorem nonzero_mul_eq_swap_mul (a b : A) : σ (a * b) \ {0} = σ (b * a) \ {0} :=
+begin
+  suffices h : ∀ (x y : A), σ (x*y) \ {0} ⊆ σ (y*x) \ {0},
```
spacing is off here (mathlib likes spaces between all operations). also I don't like the proof but I don't see a better one; I assume this isn't true if you include `0`? why isn't it?

j-loreaux replied:
ah, sorry, I thought the spacing requirement was just for statements; I'll fix that and keep it in mind for the future. Right, it's not valid if you include zero in general. I doubt there is a substantially different proof. The simple example is to take `a` to be the unilateral shift on a separable infinite dimensional Hilbert space, and `b` to be it's adjoint. Then `b*a = 1` but `a*b` is a nontrivial projection; the spectrum of the former is `{1}`, while the latter has spectrum `{0,1}`. Including `0` might be possible in *finite* C^*-algebras, but we are a ways off from even the definition of this.

j-loreaux committed:
make changes suggested by review

j-loreaux committed:
rename `resolvent` to `resolvent_set`

This is for consistency with the naming in the
mathematical community because `resolvent` is
reserved for the function which acts on the
`resolvent_set` by taking each `r : R` to the
inverse of `algebra_map R A r - a`.

dupuisf reviewed (approved):
Looks good to me!

j-loreaux committed:
Merge branch 'master' into j-loreaux/field-spectrum

j-loreaux committed:
adjust spacing to accord with mathlib style guidelines

j-loreaux committed:
fix documentation

jcommelin commented on src/algebra/algebra/spectrum.lean:
```diff
@@ -21,10 +21,13 @@ This theory will serve as the foundation for spectral theory in Banach algebras.
+* `spectrum.unit_mem_mul_iff_mem_swap_mul` and `spectrum.preimage_units_mul_eq_swap_mul`: the
+  units (of `R`) in `σ (a*b)` coincide with those in `σ (b*a)`.
+* `spectrum.scalar_eq`: in a nontrivial algebra over a field, the spectrum of a scalar is
+  a singleton.begin
```
```suggestion
  a singleton.
```

jcommelin reviewed (commented):
LGTM

j-loreaux committed:
Update src/algebra/algebra/spectrum.lean

Co-authored-by: Johan Commelin <johan@commelin.net>
